ocamllib/src/cursorinfo/cursorInfo.ml

```diff
@@ -0,0 +1,424 @@
+(* Information at the cursor *)
+open Typing
+
+type t = {
+  current_term : Syntax.t;
+  (*the currently focussed term (use to decide whether we can go down) *)
+  (*is_root: bool; (*boolean value of whether or not cursor is at root. simpler version of vv*)  *)
+  parent_term : Syntax.t option;
+  (* parent of current term (use to decide whether we can go up)  *)
+  ctx : (Var.t * int) list;
+  (* variable types *)
+  typ_ctx : (Var.t * Type.t) list;
+  (*mapping of vars in scope to types (use to determine vars in scope)    *)
+  expected_ty : Type.t option;
+  (* analyzed type of cursor_term; build up through recursion (use with ctx to determine viable insert actions) *)
+  actual_ty : Type.t option;
+      (* result of calling Syn on current_term (use to determine wrapping viability)  *)
+}
+
+(* Given a zippered AST, return the information (expr, typing, etc.) at the cursor *)
+let get_cursor_info (e : Syntax.z_t) : t =
+  let rec recurse (node : Syntax.z_t) (parent : Syntax.t option)
+      (def_cont : (Var.t * int) list) (typ_cont : Context.t)
+      (pred_type : Type.t option) (ind : int) : t =
+    let current : Syntax.t =
+      match node with
+      | ZENode zparent -> ENode (Expr.unzip_ast zparent)
+      | ZTNode zparent -> TNode (Type.unzip zparent)
+    in
+    match node with
+    | ZENode (Cursor cursed) ->
+        {
+          current_term = ENode cursed;
+          parent_term = parent;
+          ctx = def_cont;
+          typ_ctx = typ_cont;
+          expected_ty = pred_type;
+          actual_ty = synthesis typ_cont cursed;
+        }
+    | ZENode (EUnOp_L (OpNeg, argl)) ->
+        let exp_typ =
```

Comment:
the expected type should just be int here i think; negation expects an int regardless of whether the negation itself is in an appropriately-typed context